### ISSUE TITLE
[BUG] setTransactionTimeout can't be called with value larger than the upper limit of short int

### ISSUE THREAD

**Issue by xiaolaohu0826** [microsoft/mssql-jdbc#1098]
## Driver version
mssql-jdbc-7.2.2

## SQL Server version
SQL Server 2012

## Client Operating System
Microsoft Windows 10

## JAVA/JVM version
Java version 1.8.0

## Problem description
When trying to set the transaction timeout with value larger than 32767(short int upper limit) and start the transaction, the driver would crash with a weird error generated.

Please refer to pull request  #1096

1. Expected behaviour: the transaction is expected to finish successfully
2. Actual behaviour: the driver crashes and generates the following error:
3. Error message/stack trace:```
javax.transaction.xa.XAException: com.microsoft.sqlserver.jdbc.SQLServerException: The value is not set for the parameter number 0.
```
4. Any other details that can be helpful:
I guess the transaction timeout is stored as a short int in the inner structure, but still passed as a int when being prepared. please refer to my pull request.

## JDBC trace logs
<!--- Provide the JDBC trace logs. Instructions can be found here: https://docs.microsoft.com/sql/connect/jdbc/tracing-driver-operation?view=sql-server-2017 -->
Please refer to the pull request $1096

## Reproduction code
Refer to the pull request
<!--- Provide standalone JAVA code that reproduces the issue, preferably raw JDBC code. -->

**Comment by ulvii** [comment]
Hi @xiaolaohu0826 ,

Thank you for reporting the issue, we will investigate and get back to you soon.

**Comment by lilgreenbird** [comment]
hi @xiaolaohu0826,

I could repro the failure but only if I set the timeout to something > 81568 (ie > max short) and I'm seeing a different "connection reset" error. Do you see the error when setting a timeout value to say 32768?

I will continue to investigate, but I've also noticed that if I set the XAMaxTimeout registry value to a value other than 0, then I could setTransactionTimeout() up to max int value and still successfully do the xa_start.

It would really help if you could provide the code you used to repro the issue since I'm not sure why I'm seeing a different error than you. Also, could you please tell me what version of SQL Server you are running and if you have the XATimeout registry values set and if so, to what values? Please refer to [Configuring server-side timeout settings for automatic rollback of unprepared transactions](https://docs.microsoft.com/en-us/sql/connect/jdbc/understanding-xa-transactions?view=sql-server-2017#BKMK_ServerSide) for a description of the registry values.

**Comment by xiaolaohu0826** [comment]
Hi @lilgreenbird,

Thanks for investigating. I tried several values and yes, you are right. Setting the value to 32768 would not generate errors, but setting it to a number larger than 81568 would. Is 81568 a pre-defined upper limit value?

I did not see the "connection reset" error. Instead, it returned false when I tried to enlist resource in the XA transaction.

I don't have XAMaxTimeout and XADefaultTimeout under the ```HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SQL Server\MSSQL<version>.<instance_name>\XATimeout```key, and I tried to create them with values larger than 81568, it still failed.

The code that I used to reproduce this issue is a bit hard to trace back because it's used in an complicated industry project. I tried to extract out the important parts:

grab the driver -> get the XADatasource -> get the XAResource -> set the timeout -> prepare a(ny) statement -> call resource.start(xid) on it.

These should be enough to reproduce.

For the number 81568 you just mentioned. I am wondering could you please add a line to indicate if the number could be successfully set as a timeout value, for example:
```
if(seconds > 81568) return false;
```

In this way, when people try to set a timeout value to a resource and get a false returned value, they know the number exceeds the limit. Therefore, they could pick a smaller timeout value to set.

**Comment by lilgreenbird** [comment]
hi @xiaolaohu0826 ,

I'm still investigating why the 81568 value that should not be an upper limit I just found that while testing. As I mentioned previously I was able to repro the failure but didn't get the same error message as you that's why I asked for your code to see what we're doing differently?  Can you also please tell me what SQL Server version you were using? 

Also, I wanted to confirm, when you tried creating the XAMaxTimeout registry and set value to 81568 you still got the same error?

**Comment by xiaolaohu0826** [comment]
hi @lilgreenbird,

I tried to extract the minimum code from the complex codebase, and tried something like this, but still cannot reproduce the error even if I set the timeout value larger than 81568:
```package xaTransactionTest;

import java.sql.DriverManager;
import javax.sql.XAConnection;
import javax.sql.XADataSource;
import javax.transaction.xa.XAResource;
import javax.transaction.xa.Xid;
import com.microsoft.sqlserver.jdbc.SQLServerXADataSource;

/**
 * 
 */
public class XAResourceTest
{
   public static void main(String[] args) throws Exception
   {
      //DriverManager.
      SQLServerXADataSource ds = (SQLServerXADataSource)Class.forName("com.microsoft.sqlserver.jdbc.SQLServerXADataSource").newInstance();
      ds.setServerName("localhost");
      ds.setPortNumber(1433);
      ds.setDatabaseName(your_database_name);
      XAConnection connection = ds.getXAConnection(your_userid,your_password);
      XAResource xar = connection.getXAResource();

      xar.setTransactionTimeout(99999);
      try
      {
         xar.start(new XidImpl(), XAResource.TMNOFLAGS);
      }
      catch (Exception e)
      {
         int a = 0;
         return;
      }
   }

   public static class XidImpl implements Xid
   {
      /**
       * @see javax.transaction.xa.Xid#getBranchQualifier()
       */

      public byte[] getBranchQualifier()
      {
         return new byte[] {0};
      }

      public int getFormatId()
      {
         return 12345;
      }

      public byte[] getGlobalTransactionId()
      {
         return new byte[] {1,1,1,1};
      }
   }
}
```
I am sorry that I am not able to extract the minimum repro code. The error only appears when I tried to start the industry server. However, I checked the error message again and found it's the same as yours: "connection reset". I am using the SQL Server 2012.

I tried to create XAMaxTimeout and set it to 81568, and it works! With this registry being set, I can set transaction timeout to values like 86400 and it does not break.

Now the only problem we have is why would 81568 be an upper limit. If it truly is an upper limit, would you think it's good to add a line inside setTransactionTimeout() to return a boolean to determine if the incoming value is larger than that or not?

Thanks for your investigation and have a nice weekend!

**Comment by xiaolaohu0826** [comment]
Hello @lilgreenbird,

I found that if we set XAMaxTimeout registry to 10 seconds, call setTransactionTimeout(15), and sleep for 11 and the transaction rolls back, which means the registry max timeout is respected.

In this case, I think the setTransactionTimeout(15) should return false, indicating the timeout value we want to set is larger than the registry value(max timeout value), instead of true as it currently returns true for any number we passed in.

Could you please take a look on this problem too? In my opinion, it would be more intuitive to return false if the value being passed into "setTransactionTimeout" is larger than the system registry max value. Thanks!

**Comment by lilgreenbird** [comment]
hi @xiaolaohu0826,

You are correct the *XAMaxTimeout* and *XADefaultTimeout* registry values control the timeout behavior. The *XAMaxTimeout* value is max timeout that the SQL Server will allow. The timeout set via the *setTransactionTimeout* API however, is a client timeout setting. As per the documentation, if the client specifies a timeout, then "the client's timeout will be used as long as it is less than the max value". 

As per Java spec the *setTransactionTimeout* API sets the current timeout value for **this XAResource instance**, and should return true if the operation is performed successfully. The client can set this value to any value up to max integer value, and the method should return true if this is done successfully.  The server settings are specific to SQL server which is likely on a remote server or network they can be changed dynamically by the DB administrator.  They are separate from the the *XAResource* timeout set by the client. The server timeout values are read by the server when the *start* method is called, and could be different from when the client called *setTransactionTimeout* earlier.

By the way, I am currently working on a fix for the max timeout issue. Are you currently blocked by this issue i.e. do you have a use case where you need to set the transaction timeout to > 81568?

**Comment by xiaolaohu0826** [comment]
Hi @lilgreenbird ,

Not strictly blocked by this. We just want to take a larger range to ensure our transactions would be successfully committed. There are only some seeding data that may or may not take a while.

I'm glad to hear that you are working on a fix, and will just wait until you the patch is released. Thank you!

**Comment by lilgreenbird** [comment]
hi @xiaolaohu0826,
this issue will remain open until we release a new DLL containing the fix. You can subscribe to be notified when this issue gets closed. Thanks for reporting the bug to us.

**Comment by ulvii** [comment]
The issue is fixed in 8.4.0 release, closing.
